onnxruntime/core/providers/dnnl/dnnl_node_capability.cc

```diff
@@ -1075,4 +1075,95 @@ bool DnnlLayerNormalizationNodeCapability::IsDimensionSupported(const Node* node
   return true;
 }
 
+// DnnlConcatNodeCapability class
+//-------------------------------------
+
+bool DnnlConcatNodeCapability::Supported(const Node* node, const GraphViewer& graph_viewer) const {
+  ORT_UNUSED_PARAMETER(graph_viewer);
+  if (!IsTypeSupported(node)) {
+    return false;
+  }
+  if (!AreAllInputsOfSameType(node)) {
+    return false;
+  }
+  if (!AreAxisAndDimensionsSupported(node)) {
+    return false;
+  }
+  return true;
+}
+
+bool DnnlConcatNodeCapability::AreAllInputsOfSameType(const Node* node) const {
+  // All inputs must be of the same type
+  const auto node_inputs = node->InputDefs();
+  if (!node_inputs.empty() && node_inputs[0]->TypeAsProto() != nullptr) {
+    const auto node_datatype = node_inputs[0]->TypeAsProto()->tensor_type().elem_type();
+    // Ensure that any other inputs have the same datatype as the first
+    for (size_t i=1; i<node_inputs.size(); ++i) {
+      if (node_inputs[i]->TypeAsProto() != nullptr && 
+          node_inputs[i]->TypeAsProto()->tensor_type().elem_type() != node_datatype) {
+        return false;
+      }
+    }
+  }
+  return true;
+}  
+
+bool DnnlConcatNodeCapability::AreAxisAndDimensionsSupported(const Node* node) const {
+  const auto& attributes = node->GetAttributes();
+  auto axis_attr = attributes.find("axis");
+  if (axis_attr != attributes.end()) {
+    if (axis_attr->second().type() != ::ONNX_NAMESPACE::AttributeProto_AttributeType::AttributeProto_AttributeType_INT) {
+      // Axis must be an integer
+      return false;
+    }
+  } else {
+    // Axis is required.
+    return false;
+  }    
+
+  const auto& node_inputs = node->InputDefs();
+
+  // Validate axis
```

Comment:
I feel uneasy about verifying what appears to be OpSet conformance below. I assume that a false value will fallback to the CPU EP (which will reject it at runtime?). Does removing these feel reasonable?

Reply:
You should not have to check something that is part of the op specification.  Typically the only things you need to check are things that could run on the CPU but would fail when run on oneDNN ep.

Returning false value will let other eps claim support.  This is almost always results in the the CPU ep executing the operator.